### PR TITLE
Add metadata query parameter on PUT

### PR DESCRIPTION
**Pull request by riya-17**
This PR will allow the Agent (or other client) to define metadata keys when initially uploading a dataset to the Pbench Server by specifying the ?metadata query parameter on the PUT.

This PR is based on #3187 and should be merged after that.

### CODE REVIEW

**Comment by portante** [comment]
Legacy agent unit tests failing due to new command line option, `--metadata`:

```
test-23 pbench-results-move --help
*** /home/jenkins/pbench/agent/util-scripts/gold/pbench-results-move/test-23.txt	2023-02-10 20:26:04.853670162 +0000
--- /var/tmp/pbench-test-utils/output.txt	2023-02-10 20:27:46.167754956 +0000
***************
*** 12,17 ****
--- 12,18 ----
    --controller TEXT              Override the default controller name
    --delete / --no-delete         Remove local data after successful copy
                                   [default: delete]
+   -m, --metadata TEXT            metadata to add information for server
    --show-server TEXT             Display information about the pbench server
                                   where the result(s) will be moved (Not
                                   implemented)
FAIL - test-23
test-24 pbench-results-push --help
*** /home/jenkins/pbench/agent/util-scripts/gold/pbench-results-push/test-24.txt	2023-02-10 20:26:04.853670162 +0000
--- /var/tmp/pbench-test-utils/output.txt	2023-02-10 20:27:46.430752580 +0000
***************
*** 11,16 ****
--- 11,17 ----
    -C, --config PATH              Path to a pbench-agent configuration file
                                   (defaults to the '_PBENCH_AGENT_CONFIG'
                                   environment variable, if defined)  [required]
+   -m, --metadata TEXT            metadata to add information for server
    --token TEXT                   pbench server authentication token
                                   [required]
    --help                         Show this message and exit.
FAIL - test-24
```

File: lib/pbench/cli/agent/commands/results/results_options.py
Comment (npalaska) [inline lib/pbench/cli/agent/commands/results/results_options.py:24]: Do we want to expand a little more on this help? Maybe something like:
```
Optional flag to add metadata to the dataset, 
the server may take additional actions based on the value.
```
Should we add an example to this help like `server.archiveonly:y` would mark the dataset for archival only?

File: lib/pbench/test/unit/agent/task/test_copy_result_tb.py
Comment (npalaska) [inline lib/pbench/test/unit/agent/task/test_copy_result_tb.py:127]: Metadata and access default values are `None` so I guess we don't need an `if-else` block. We can just call `crt.copy_result_tb("token", access, metadata)`

**Review by npalaska — DISMISSED** [review]
Looks good overall, I just have minor comments.

**Comment by npalaska** [comment]
Added do not merge since this one is dependent on #3187

File: lib/pbench/test/unit/agent/task/test_copy_result_tb.py
Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:127]: I could go either way on that:  while simpler code is generally nicer, having the caller provide the `None` value is slightly different from having the callee provide it by default -- it's true that, as the code is presently constructed, the result is the same, but, if the CUT were to be changed (so that `None` is still acceptable but no longer the default), it would be good if the unit test exercised both cases.  (And, in general, we shouldn't be changing the unit tests when we change the CUT.)

File: lib/pbench/test/unit/agent/task/test_results_move.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/task/test_results_move.py:122]: Re line 122, it looks like, based on line 121, the expected `exit_code` value is **_2_**, and that's not a "success".

File: lib/pbench/test/unit/agent/task/test_results_move.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/task/test_results_move.py:123]: It seems like this assertion should be more specific than just, "Did we get something on `stderr`?"  And, the token "Usage" seems weird in this context -- it seems like the message should mention that `stderr` is unexpectedly empty.

File: lib/pbench/test/unit/agent/task/test_results_push.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/task/test_results_push.py:116]: Rather than using `find()` with a comparison against `-1`, how about using `in`?
```
        assert (
            "Invalid value for 'RESULT_TB_NAME': "
            "File 'pbench.sat:FOO' does not exist."
            in result.stderr
        )
```
(If you get complaints about mismatches between bytes-like objects and strings, I think there is an accessor for `result` which returns `stderr` a proper string, or you can always declare the comparison to be a bytes literal by putting a `b` in front of the opening quote.)

File: lib/pbench/test/unit/agent/task/test_results_push.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/task/test_results_push.py:105]: What happens if we use this combination?
```
                TestResultsPush.META_SWITCH,
                TestResultsPush.META_TEXT_TEST + ',' + TestResultsPush.META_TEXT_FOO,
```
(i.e., `-m pbench.sat:FOO,pbench.test:TEST`).  Does that work as expected?

(I think that would be a better test than this one, which effectively provides two files rather than an orphaned metadatum.)

**Review by webbnh — CHANGES_REQUESTED** [review]
Most of my comments are relatively small things, but the assertion message in `test_results_move.py` should be fixed, and it would be good to improve the help text for the `--metadata` option.

File: lib/pbench/cli/agent/commands/results/push.py
Comment (riya-17) [reply lib/pbench/cli/agent/commands/results/push.py:68]: yes, I can do that now, earlier I was using a different approach.

File: lib/pbench/test/unit/agent/task/test_results_push.py
Comment (riya-17) [reply lib/pbench/test/unit/agent/task/test_results_push.py:105]: It is not throwing any errors but I would love to experiment that in real environment.

**Review by webbnh — DISMISSED** [review]
I'm approving this, but there are still a few outstanding items which I think should be addressed before this is merged:

- [ ] I believe that this PR depends on #3220; assuming that the removal of `get_path_exists_mock()` from `test_copy_result_tb.py` is rolled back in that PR, this PR should be using that function to set up the `Path.exists()` mock in `test_with_metadata()` ([ref](https://github.com/distributed-system-analysis/pbench/pull/3245#discussion_r1103491898))
- [x] I think there is a redundant assertion that should be removed ([ref](https://github.com/distributed-system-analysis/pbench/pull/3245#discussion_r1103494032), and below)
- [ ] I have a small but important wording suggestion (see below)

File: lib/pbench/cli/agent/commands/results/results_options.py
Comment (dbutenhof) [reply lib/pbench/cli/agent/commands/results/results_options.py:28]: "May" would be good. I decided not to comment on the over-simplified help here; but, technically, the server is less constrained here. `?metadata=server.origin:foo&metadata=server.archiveonly:t` is handled *exactly* the same as `?metadata=server.origin:foo,server.archiveonly:t`.

I really don't want to overcomplicate the help on the agent side by getting into that ... but if the user knows, "may" loosely avoids suggesting that the agent is imposing some additional formal constraint. 😁

**Review by dbutenhof — APPROVED** [review]

**Review by npalaska — APPROVED** [review]
Looks good now.

**Comment by portante** [comment]
Moved to `Draft` mode until PR #3187 lands.

**Comment by riya-17** [comment]
> Moved to `Draft` mode until PR #3187 lands.

PR #3187 is closed

File: lib/pbench/test/unit/agent/task/test_results_move.py
Comment (riya-17) [reply lib/pbench/test/unit/agent/task/test_results_move.py:124]: No, I was not planning to remove l.123

**Comment by dbutenhof** [comment]
> > Moved to `Draft` mode until PR #3187 lands.
> 
> PR #3187 is closed

So, would you like to move it out of Draft mode? 😁

File: lib/pbench/test/unit/agent/task/test_results_move.py
Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_results_move.py]: OK, then perhaps you might want to tweak it some:  if the assertion fails, then `result.stderr` is probably an empty string, so the output will be
```
Usage: '', stdout: '<...stuff...>'
```
which doesn't really convey much to me in terms of indicating why the test failed.

On the other hand, if you simply remove line 123, then, in the situation where it would have failed, we'll see the failure of the assertion at line 124, in which PyTest will helpfully display the fact that `result.stderr` is empty (not to mention the fact that the assertion expression in line 124 is pretty expressive -- i.e., it shows what we expected, and PyTest will show what we actually got).

**Review by webbnh — DISMISSED** [review]
Looks good, but I would still like to see [`get_path_exists_mock()` used](https://github.com/distributed-system-analysis/pbench/pull/3245#discussion_r1103491898).

**Comment by riya-17** [comment]
> Looks good, but I would still like to see [`get_path_exists_mock()` used](https://github.com/distributed-system-analysis/pbench/pull/3245#discussion_r1103491898).

I am not sure why it shows like that but I did restore it. will do it again.

**Review by dbutenhof — DISMISSED** [review]
A discussion for another day and PR, but given the history of this mock, (and many other similar incidents in the past) I sometimes think we should develop a policy of carefully isolating all python general library mocks inside a `monkeypatch` context manager with `assert` statements safely *outside* to make sure nothing confuses `pytest`.

**Review by portante — APPROVED** [review]

**Review by portante — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
Ship it!  😃

**Pull request merged**
